### DISCUSSION TITLE
IBinding is nolonger user client implementable

### DISCUSSION THREAD

**Discussion by mohdimas** [AvaloniaUI/Avalonia#13451] [Q&A]
Hi there,

We are upgrading from Avalonia **0.10** to **0.11**. We have several classes that is deriving from `IBinding`. But now the class has been marked as `NotClientImplementable`, so is there any alternative to it?

Many thanks

**Comment by thevortexcloud** [comment]
My guess is that you are meant to use `Binding` or `BindingBase` (which implements `IBinding`) now. A lot of interfaces got removed for various reasons. But one of the Avalonia core members can probably answer better than me.

**Reply by mohdimas** [reply]
Thanks for the reply. We need to implement/override the `IBinding.Initiate` in our derived class. But just setting the parent class to `Binding` or `BindingBase` won't allow us to do that.

**Reply by maxkatz6** [reply]
Can you be more specific? Why do you need to override Initiate? What should your IBinding do that typical BindingBase doesn't?

**Reply by mohdimas** [reply]
we have several classes which derive from `IBinding`. One of the use cases is to find a resource from a bindable `Key` property.

**Comment by rabbitism** [comment]
You can check this. 

https://github.com/AvaloniaUI/Avalonia/wiki/Using-private-apis-in-nuget-packages

**Reply by mohdimas** [reply]
Many thanks 🙏

**Reply by maxkatz6** [reply]
Please don't use private APIs unless you are building platform/rendering backends or unless you need it for a **short time** during migration.

**Comment by WenLiHou** [comment]
`<TextBlock  Text="{localize:Translate {Binding Path=Text}}">`
If I can't use IBinding, how can I implement a TranslateExtension like the one above? The TranslateExtension receives the key or key binding, and returns the localized string by looking it up in the LocalizationManager, and the LocalizationManager implements the INotifyPropertyChanged interface.

**Reply by thevortexcloud** [reply]
You might be able to use a markup extension that takes a binding as the argument. But I am not sure if that will work for on the fly updates.

**Reply by WenLiHou** [reply]
I've implemented this extension and it worked fine in previous versions of avalonia, but after I updated to the latest nightlybuilds, it prompted that the IBinding implementation was not supported.

**Reply by thevortexcloud** [reply]
You don't need to inherit from `IBinding` to make a markup extension. The base markup extension is public.

https://github.com/AvaloniaUI/Avalonia/blob/e35e94902eb4eef415e205fa0ca7e911ac2aa2f4/src/Markup/Avalonia.Markup.Xaml/MarkupExtension.cs#L5

As is the reflection binding extension:

https://github.com/AvaloniaUI/Avalonia/blob/e35e94902eb4eef415e205fa0ca7e911ac2aa2f4/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/ReflectionBindingExtension.cs#L12

**Comment by timunie** [comment]
Try something like this: 

```cs
public enum LocaleKey
{
  Yes,
  No,
  ...
}

public class Localize : MarkupExtension
{
    private LocaleKey Key { get; }
    
    public Localize(LocaleKey key)
    {
        Key = key;
    }
    
    public override object ProvideValue(IServiceProvider serviceProvider)
    {
        return new Binding($"[{Key}]")
        {
            Source = LocaleManager.Instance // this is the translator I use
        };
    }
}
```

> [!NOTE] 
> It returns a reflection binding. Creating a CompiledBinding should also work, but needs much more work. This is up to you if really needed. We have some Q&As discussing it.

**Comment by rabbitism** [comment]
To add one, Avalonia actually use duck typing for mark up extension, as long as you have a `ProvideValue` method, the compiler will consider it as a valid markup extension.

**Comment by WenLiHou** [comment]
I tried to change it to:
```C#
  //return binding;
  if (Name is string str)
  {
      var localization = ((App)Application.Current!).Services.GetRequiredService<LocalizationResourceManager>();

      return new FloatingResourceObservable(localization!, str).ToBinding();
  }
  else if (Name is CompiledBindingExtension extension)
  {
      return extension;
  }
  else
  {
      return AvaloniaProperty.UnsetValue;
  }
```
it's ok.
